server/src/main/java/org/elasticsearch/index/mapper/DocumentParserContext.java

```diff
@@ -131,6 +129,7 @@ private DocumentParserContext(
         this.dynamic = dynamic;
         this.fieldsAppliedFromTemplates = fieldsAppliedFromTemplates;
         this.copyToFields = copyToFields;
+        this.totalFieldsCountBeforeUpdate = mappingLookup.getTotalFieldsCount();
```

Comment:
This isn't really the cheapest of methods. If we want to use the field count, then we should only compute it in case we need it. We shouldn't be doing this for every document we parse?

Before we get into that, why do we need this and aren't using the mapping version id as we discussed last week, was there an issue with that?

Reply:
> Before we get into that, why do we need this and aren't using the mapping version id as we discussed last week, was there an issue with that?

I tried to use the mapping version but I found an issue with it. Apparently, there are two cluster state updates when updating the mapping. There's one that updates just the version and another one that updates the fields. What was happening is that for the same mapping version, I was seeing different field counts. This lead to `DocumentParserContext#addDynamicMapper` adding fields thinking that the number of fields hasn't been exceeded. When checking in `TransportShardBulkAction#executeBulkItemRequest`, the version was still the same but the number of fields may have been increased, making the mapping update fail due to exceeding the limit. But since the mapping version hasn't changed, we wouldn't re-try the request, returning an error to the client. Therefore, `DynamicMappingIT#testConcurrentDynamicIgnoreBeyondLimitUpdates` was failing occasionally.

> This isn't really the cheapest of methods. If we want to use the field count, then we should only compute it in case we need it. We shouldn't be doing this for every document we parse?

Inside of `DocumentParserContext#addDynamicMapper`, we're already calling `mappingLookup.getTotalFieldsCount()` via `MappingLookup#checkFieldLimit`. I also don't understand why this method isn't cheap, as it just adds a couple of integers:

https://github.com/elastic/elasticsearch/blob/406a1897c6f126a70f23e341c737304526c41485/server/src/main/java/org/elasticsearch/index/mapper/MappingLookup.java#L225-L227

Another benefit of checking the field count rather than the mapping version is that we should be doing fewer false-positive mapping update retries that aren't likely to resolve the issue. We're not retrying on any concurrent changes to the mapping, just concurrent updates that have added a field.

Another thing that we could try is to not read the fields count in `DocumentParserContext` is to modify the retry condition in TransportShardBulkAction#onMappingUpdateFailure` to something like this:

```
if (mappingLookup.exceedsLimit(indexSettings.getMappingTotalFieldsLimit(), 1) && context.isMappingUpdateRetry() == false)
```

However, that would lead to re-tries even if where were no concurrent updates to the mapping, once the field limit is hit.

We could also combine both approaches but maybe that's becoming unnecessarily complex. I do like the solution of checking whether a field has been added concurrently. I don't quite understand why getting the total field count is expensive.

Reply:
>  Apparently, there are two cluster state updates when updating the mapping. There's one that updates just the version and another one that updates the fields.

No that shouldn't be true. There's only a single cluster state update here. If we see that kind of conflict we need to investigate and fix it. Looking into that ... 

>  I don't quite understand why getting the total field count is expensive.

Maybe I was wrong here sorry about that, I mixed up the method on the mapping lookup and the new method for counting the dynamic fields that was a nested loop. My bad :) The fact that the mapping version isn't properly unique here is troubling nonetheless, we can't keep adding complexity to this code when assumptions about consistency are violated already, I'll hurry on this :)